internal/campaigns/workspace.go

```diff
@@ -0,0 +1,94 @@
+package campaigns
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/sourcegraph/src-cli/internal/api"
+	"github.com/sourcegraph/src-cli/internal/campaigns/graphql"
+)
+
+// WorkspaceCreator implementations are used to create workspaces, which manage
+// per-changeset persistent storage when executing campaign steps and are
+// responsible for ultimately generating a diff.
+type WorkspaceCreator interface {
+	// Create should clone the given repository, and perform whatever other
+	// initial setup is required.
+	Create(context.Context, *graphql.Repository) (Workspace, error)
+
+	// DockerImages returns any Docker images required to use workspaces created
+	// by this creator.
+	DockerImages() []string
+}
+
+// Workspace implementations manage per-changeset storage when executing
+// campaign step.
+type Workspace interface {
+	// Prepare is called once, immediately before the first step is executed.
+	// Generally, this would perform whatever Git or other VCS setup is required
+	// to establish a base upon which to calculate changes later.
+	Prepare(ctx context.Context) error
```

Comment:
This is the only one I'm unsure about: since we do have a `WorkspaceCreator` and _specific implementations_: is there a reason why we shouldn't the things that are being done in `Prepare` into the `Create` method of the corresponding workspace creators?

I think that could be nice, because it would mean that (1) `WorkspaceCreator` creates a ready-to-go workspace and (2) the `Workspace` interface becomes smaller and easier to use (you don't have to manipulate a workspace before you can use it)

Reply:
I went back and forth on this: originally, this survived because I hadn't fully fleshed out the creator interface and was trying to avoid changing too much about how the bind workspace operated (since it had an analogue for `Prepare`), but in the end, I left it mostly out of inertia and a vague feeling that there might be scenarios where it makes sense to keep these separate.

But, realistically, this is used in one place right now, immediately after creating the workspace. So I think you're right, and I'll change it.